connection.go

```diff
@@ -671,41 +675,86 @@ func (c *Connection) readFrames(_ uint32) {
 			return
 		}
 
-		// call req and call res messages may not want the frame released immediately.
-		releaseFrame := true
-		switch frame.Header.messageType {
-		case messageTypeCallReq:
-			releaseFrame = c.handleCallReq(frame)
-		case messageTypeCallReqContinue:
-			releaseFrame = c.handleCallReqContinue(frame)
-		case messageTypeCallRes:
-			releaseFrame = c.handleCallRes(frame)
-		case messageTypeCallResContinue:
-			releaseFrame = c.handleCallResContinue(frame)
-		case messageTypeInitReq:
-			c.handleInitReq(frame)
-		case messageTypeInitRes:
-			releaseFrame = c.handleInitRes(frame)
-		case messageTypePingReq:
-			c.handlePingReq(frame)
-		case messageTypePingRes:
-			releaseFrame = c.handlePingRes(frame)
-		case messageTypeError:
-			releaseFrame = c.handleError(frame)
-		default:
-			// TODO(mmihic): Log and close connection with protocol error
-			c.log.WithFields(
-				LogField{"header", frame.Header},
-				LogField{"remotePeer", c.remotePeerInfo},
-			).Error("Received unexpected frame.")
+		var releaseFrame bool
+		if c.relay == nil {
+			releaseFrame = c.readFrameNoRelay(frame)
+		} else {
+			releaseFrame = c.readFrameRelay(frame)
 		}
-
 		if releaseFrame {
 			c.framePool.Release(frame)
 		}
 	}
 }
 
+func (c *Connection) readFrameRelay(frame *Frame) bool {
+	releaseFrame := true
+
+	// call req and call res messages may not want the frame released immediately.
+	switch frame.Header.messageType {
+	case messageTypeCallReq, messageTypeCallReqContinue, messageTypeCallRes, messageTypeCallResContinue:
+		if err := c.relay.Relay(frame); err != nil {
+			c.log.WithFields(
+				LogField{"header", frame.Header},
+				LogField{"remotePeer", c.remotePeerInfo},
+			).Error("Failed to relay frame.")
+		}
+		releaseFrame = false
+	case messageTypeInitReq:
```

Comment:
what do you think about having a `default` that just calls `handleFrameNoRelay` here instead of duplicating these calls